File: 백준/Silver/2108. 통계학/통계학.cc
// 2108 통계학

#include<iostream>
#include<algorithm>
#include<cmath>
#include<stdio.h>
using namespace std;

int average(int* num, int n){
    double sum = 0;
    for(int i = 0; i<n; i++){
        sum += num[i];
    }
    return round(sum/n);
}

int median(int* num, int n){
    sort(num,num+n);
    return num[n/2];
}

int mode(int* num, int n){
    int count[8001]={0};
    int max=0;                          // max Count
    int maxIndex=0;                   // max data Index 0~8000


    for(int i = 0; i<n; i++){
        count[num[i]+4000]++;
    }


    for(int j = 0; j<8001; j++){
        if(count[j] > max){
            max = count[j];
            maxIndex = j;
        }      
    }


    int result = maxIndex;

    for(int i = maxIndex+1; i<8001; i++){
        if(count[i] == max){
            result = i;
            break;
        }
    }

    return result-4000;

}

int range(int* num, int n){
    sort(num,num+n);
    int result = num[n-1]-num[0];
    return result;
}

int main(){
    int n;
    scanf("%d", &n);

    int num[500000];

    for(int i = 0; i<n; i++){
        scanf("%d", &num[i]);
    }

    printf("%d\n",average(num,n));
    printf("%d\n",median(num,n));
    printf("%d\n",mode(num,n));
    printf("%d\n",range(num,n));


}

File: 백준/Silver/2108. 통계학/통계학.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args)throws IOException {
        // System.out.println((double)((int)(3.001*100)/100.00));
        // System.out.printf("%.2f\n",(double)(int)(3.1462*100)/100);
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        int[] nums = new int[n];
        int sum = 0;
        int [] count = new int[8001];


        for(int i = 0; i<n; i++){
            int temp = Integer.parseInt(br.readLine());
            nums[i] = temp;
            sum += temp;
            count[temp+4000]++;
        }

        Arrays.sort(nums);

        int maxIndex = 0;
        int maxValue = 0;
        int maxFreq = 0;

        for(int i = 0; i<count.length; i++){
            if(count[i] > maxValue){
                maxIndex = i;
                maxValue = count[i];
                maxFreq = i-4000;
            }
        }


        
        for(int i = maxIndex+1; i<count.length; i++){
            if(count[i] == maxValue){
                maxFreq = i-4000;
                break;
            }
        }


        
        int range = nums[nums.length - 1] - nums[0];

        System.out.println(Math.round((double)sum/n));
        System.out.println(nums[n/2]);
        System.out.println(maxFreq);
        System.out.println(range);


    }
    
}